### PR TITLE
Constext

### PR DESCRIPTION
**Pull request by MarikaBBB**
The was an excess of white space beneath the footer on the shopping cart page. 

Adjusted Flex Layout:
- Applied flex-grow to the ShopCart component to allow it to expand and fill the available space between the header and footer.
Ensured the footer stays at the bottom of the viewport by leveraging a flex layout in App.tsx.

Now the ShopCart section grows naturally, filling the available space, keeping the footer in place.
Testing & Validation:
Screenshot 2024-10-23 at 23 04 19

### CODE REVIEW

**Comment by MarikaBBB** [comment]
- IconWithText Component:
Refactored the component to use Link from react-router-dom for proper navigation handling.

- Header Component:
The Account icon now navigates to the /UserLogin route, directing users to the login page.
The Cart icon now navigates to the /ShopCart route, leading users to their shopping cart.

File: src/App.css [hunk 1]
```diff
@@ -1,42 +1,16 @@
...
+    min-height: 100vh; 
   }
-}
-
-@media (prefers-reduced-motion: no-preference) {
-  a:nth-of-type(2) .logo {
-    animation: logo-spin infinite 20s linear;
+  
+  main {
+    flex-grow: 1;
```
Comment (maxitect) [inline src/App.css:14]: Isn't this already defined in the tailwind class inside App.tsx? Why are we redefining it here too?

File: src/App.tsx [hunk 2]
```diff
@@ -20,19 +20,19 @@ const App = () => {
...
-    </Router>
+      <Router>
+        <Header cartCount={cartCount} />
+        <main className="flex-grow">
+          <Routes>
+            <Route path="/" element={<Home setCartCount={setCartCount} />} />
+            <Route path="/UserLogin" element={<UserLogin />} />
+            <Route path="/UserSignUp" element={<UserSignUp />} />
+            <Route path="/ShopCart" element={<ShopCart setCartCount={setCartCount} />} />
+            {/* Add other routes here */}
```
Comment (maxitect) [inline src/App.tsx:31]:
```suggestion
```

File: src/Context/CartContext.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+
+interface Product {
+  id: number;
+  title: string;
+  artist: string;
+  price: number;
+  image_url: string;
+  quantity: number;
+}
```
Comment (maxitect) [inline src/Context/CartContext.tsx:11]:
```suggestion
import { Vinyl } from '../utils/types.ts';
```
This interface should be imported from the types @JoshCodedit added for all database schemas

File: src/Context/CartContext.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    if (savedCart) {
+      const items = JSON.parse(savedCart);
+      setCartItems(items);
+      calculateTotal(items);
+      const newCartCount = items.reduce((acc: number, item: Product) => acc + item.quantity, 0);
+      setCartCount(newCartCount);
+    }
+  }, [setCartCount]);
+
+  const calculateTotal = (items: Product[]) => {
```
Comment (maxitect) [inline src/Context/CartContext.tsx:51]:
```suggestion
  const calculateTotal = (items: Vinyl[]) => {
```

File: src/Context/CartContext.tsx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      .map((item) => {
+        if (item.id === productId) {
+          if (item.quantity > 1) {
+            return { ...item, quantity: item.quantity - 1 };
+          }
+          return null; // Remove item if quantity is less than 1
+        }
+        return item;
+      })
+      .filter((item): item is Product => item !== null);
```
Comment (maxitect) [inline src/Context/CartContext.tsx:74]:
```suggestion
      .filter((item): item is Vinyl => item !== null);
```

File: src/Context/CartContext.tsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  title: string;
+  artist: string;
+  price: number;
+  image_url: string;
+  quantity: number;
+}
+
+
+interface CartContextType {
+  cartItems: Product[];
```
Comment (maxitect) [inline src/Context/CartContext.tsx:15]:
```suggestion
  cartItems: Vinyl[];
```

File: src/Context/CartContext.tsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  cartCount: number;
+  totalPrice: number;
+  addToCart: (product: Product) => void;
```
Comment (maxitect) [inline src/Context/CartContext.tsx:18]:
```suggestion
  addToCart: (vinyl: Vinyl) => void;
```

File: src/Context/CartContext.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  const context = useContext(CartContext);
+  if (!context) {
+    throw new Error('useCart must be used within a CartProvider');
+  }
+  return context;
+};
+
+// taking from the ShopCArT
+const ShopCart = ({ setCartCount }: ShopCartProps) => {
+  const [cartItems, setCartItems] = useState<Product[]>([]);
```
Comment (maxitect) [inline src/Context/CartContext.tsx:37]:
```suggestion
  const [cartItems, setCartItems] = useState<Vinyl[]>([]);
```

File: src/Context/CartContext.tsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  const [totalPrice, setTotalPrice] = useState(0);
+
+  useEffect(() => {
+    const savedCart = localStorage.getItem('shoppingCart');
+    if (savedCart) {
+      const items = JSON.parse(savedCart);
+      setCartItems(items);
+      calculateTotal(items);
+      const newCartCount = items.reduce((acc: number, item: Product) => acc + item.quantity, 0);
```
Comment (maxitect) [inline src/Context/CartContext.tsx:46]:
```suggestion
      const newCartCount = items.reduce((acc: number, item: Vinyl) => acc + item.quantity, 0);
```

File: src/Context/CartContext.tsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      setCartCount(newCartCount);
+    }
+  }, [setCartCount]);
+
+  const calculateTotal = (items: Product[]) => {
+    const total = items.reduce((acc, item) => acc + item.price * item.quantity, 0);
+    setTotalPrice(total);
+  };
+
+  const increaseQuantity = (productId: number) => {
```
Comment (maxitect) [inline src/Context/CartContext.tsx:56]:
```suggestion
  const increaseQuantity = (vinylId: number) => {
```

File: src/Context/CartContext.tsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    const updatedCart = cartItems.map((item) =>
+      item.id === productId ? { ...item, quantity: item.quantity + 1 } : item
```
Comment (maxitect) [inline src/Context/CartContext.tsx:58]:
```suggestion
      item.id === vinylId ? { ...item, quantity: item.quantity + 1 } : item
```

File: src/Context/CartContext.tsx [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    );
+    updateCartState(updatedCart);
+  };
+
+  const decreaseQuantity = (productId: number) => {
```
Comment (maxitect) [inline src/Context/CartContext.tsx:63]:
```suggestion
  const decreaseQuantity = (vinylId: number) => {
```

File: src/Context/CartContext.tsx [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    const updatedCart = cartItems
+      .map((item) => {
+        if (item.id === productId) {
```
Comment (maxitect) [inline src/Context/CartContext.tsx:66]:
```suggestion
        if (item.id === vinylId) {
```

File: src/Context/CartContext.tsx [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+          }
+          return null; // Remove item if quantity is less than 1
+        }
+        return item;
+      })
+      .filter((item): item is Product => item !== null);
+    updateCartState(updatedCart);
+  };
+
+  const updateCartState = (updatedCart: Product[]) => {
```
Comment (maxitect) [inline src/Context/CartContext.tsx:78]:
```suggestion
  const updateCartState = (updatedCart: Vinyl[]) => {
```

File: src/pages/ShopCart.tsx [hunk 15]
```diff
@@ -1,95 +1,57 @@
...
+        className="bg-background-light p-4 shadow rounded-t-lg flex-grow"
+      >
+        <h1 className="text-4xl font-bold mb-4 text-text-primary text-center">Shopping Cart</h1>
+      
+        <div id="container" 
+        className="max-w-6xl mx-auto ">
+          {cartItems.length === 0 ? (
+            <p className="text-text-primary text-center">Your cart is empty</p>
+          ) : (
+            cartItems.map((product) => (
```
Comment (maxitect) [inline src/pages/ShopCart.tsx:21]:
```suggestion
            cartItems.map((vinyl) => (
```

File: src/pages/ShopCart.tsx [hunk 16]
```diff
@@ -1,95 +1,57 @@
...
-            <h2>Total Price: £{totalPrice.toFixed(2)}</h2>
-          </div>
-        </section>
-      );
-    };
- 
+                <Quantifier
+                  quantity={product.quantity}
+                  onIncrease={() => increaseQuantity(product.id)}
+                  onDecrease={() => decreaseQuantity(product.id)}
```
Comment (maxitect) [inline src/pages/ShopCart.tsx:40]:
```suggestion
                  quantity={vinyl.quantity}
                  onIncrease={() => increaseQuantity(vinyl.id)}
                  onDecrease={() => decreaseQuantity(vinyl.id)}
```

File: src/pages/ShopCart.tsx [hunk 17]
```diff
@@ -1,95 +1,57 @@
...
+              >
+                <img 
+                  src={product.image_url} 
+                  alt={product.title} 
+                  className="w-24 h-24 object-cover rounded-lg"
+                />
+                <div className="ml-4 flex-grow">
+                  <h3 className="text-lg font-medium text-text-primary">{product.title}</h3>
+                  <p className="text-text-secondary">{product.artist}</p>
+                  <p className="font-bold text-dark">£{product.price.toFixed(2)}</p>
```
Comment (maxitect) [inline src/pages/ShopCart.tsx:35]:
```suggestion
                id="vinyl"
                className="flex items-center justify-between p-4 mb-4 border border-gray-300 rounded-md bg-background-default shadow-sm"
                key={vinyl.id}
              >
                <img 
                  src={vinyl.image_url} 
                  alt={vinyl.title} 
                  className="w-24 h-24 object-cover rounded-lg"
                />
                <div className="ml-4 flex-grow">
                  <h3 className="text-lg font-medium text-text-primary">{vinyl.title}</h3>
                  <p className="text-text-secondary">{vinyl.artist}</p>
                  <p className="font-bold text-dark">£{vinyl.price.toFixed(2)}</p>
```

**Review by maxitect — CHANGES_REQUESTED** [review]
Sorry may have gone overboard with comments about product -> vinyl but it stems from using the interface inside utils/types.ts which reflects what is in the database that @JoshCodedit is managing

**Pull request closed without merging**